torch/csrc/api/include/torch/data/samplers/base.h

```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <ATen/optional.h>
+
+#include <cstddef>
+#include <vector>
+
+namespace torch {
+namespace data {
+namespace samplers {
+
+/// A `Sampler` is an object that yields indices with which to index into a
+/// dataset.
+class Sampler {
```

Comment:
FWIW In Python we do have concepts of `Sampler`s which work on single examples, and then have a `BatchSampler`, which builds up lists of indices. I guess you don't want to do it because you want the whole system to work in batches, to simplify the templates?

Reply:
Yes exactly, the design here is more batch-first